Replace new-command.md with skills (#2709)
microsoft/mcp: .github/skills/add-azure-mcp-tools/SKILL.md

│   ├── Services/
│   │   ├── I{Toolset}Service.cs
│   │   └── {Toolset}Service.cs
│   └── Models/
└── tests/
    ├── Azure.Mcp.Tools.{Toolset}.Tests/
    │   └── Azure.Mcp.Tools.{Toolset}.Tests.csproj
    ├── test-resources.bicep          (Azure service commands only)
    └── test-resources-post.ps1       (Azure service commands only)
```

Required setup steps:

1. Add package version to `Directory.Packages.props` (if Azure SDK needed)
2. Add project to `Microsoft.Mcp.slnx` and `Azure.Mcp.Server.slnx`
3. Register the new toolset in `servers/Azure.Mcp.Server/src/Program.cs` `RegisterAreas()` (alphabetical order)
4. For the primary pattern, have commands inherit directly from `SubscriptionCommand<TOptions, TResult>` and inject `ISubscriptionResolver`. Only add a shared base command if you have real cross-command logic.
5. Register both the service **and the command** as singletons in `{Toolset}Setup.cs` `ConfigureServices`:
   ```csharp
   public void ConfigureServices(IServiceCollection services)
   {
       services.AddSingleton<I{Toolset}Service, {Toolset}Service>();
       services.AddSingleton<{Resource}{Operation}Command>();
   }
   ```

> **GATE:** `dotnet build servers/Azure.Mcp.Server/src` must pass.

Reviewer: This is tied to Azure MCP tools specifically, should call this skill `add-azure-mcp-tools`. In the future we can rework this to support non-Azure tools with a shared common set of requirements and mix-in Azure, Fabric, etc specifics later.

Reviewer 2: Will rename this in a follow up PR, otherwise will lose all the PR review comments on this file and it will be a disruptive change to update all references

Reviewer 2: @copilot Open an issue to track the rename of the skills file to `add-azure-mcp-tools` and update all the references

                ["resource-group"] = ResourceGroupName,
            });

        Assert.NotNull(result);
        var items = result.Value.AssertProperty("items");
        Assert.Equal(JsonValueKind.Array, items.ValueKind);
    }
}
```

### 3c. Record and Verify

```powershell
# Record
$settings = @{ TestMode = "Record"; SubscriptionId = "<sub>"; TenantId = "<tenant>";
               ResourceGroupName = "<rg>"; ResourceBaseName = "<base>" } | ConvertTo-Json
$settings | Set-Content "tools\Azure.Mcp.Tools.{Toolset}\tests\Azure.Mcp.Tools.{Toolset}.Tests\.testsettings.json"

dotnet test tools\Azure.Mcp.Tools.{Toolset}\tests\Azure.Mcp.Tools.{Toolset}.Tests `
  --filter "FullyQualifiedName~{Resource}{Operation}"

# Push recordings
.proxy\Azure.Sdk.Tools.TestProxy push `
  -a tools\Azure.Mcp.Tools.{Toolset}\tests\Azure.Mcp.Tools.{Toolset}.Tests\assets.json

# Verify playback
# Change TestMode to "Playback" in .testsettings.json, then re-run tests
```

Create `assets.json` if it doesn't exist:
```json
{
  "AssetsRepo": "Azure/azure-sdk-assets",
  "AssetsRepoPrefixPath": "",
  "TagPrefix": "Azure.Mcp.Tools.{Toolset}.Tests",
  "Tag": ""
}
```

> ### 3c-1. Recorded Test Pitfalls

Reviewer: Is `3c-1` something specific for Skills? If not, just call this section `3d` and update the subsequent `3*` sections below to the next letter. Or just remove this header and keep what's after it in `3c`

Reviewer 2: It's kind of best practices/ common pitfalls/ things to watch out for while recording tests which is 3c.

```

#### Credential type in `.testsettings.json`

`Deploy-TestResources.ps1` sets `AZURE_TOKEN_CREDENTIALS=AzurePowerShellCredential`. If the MCP server subprocess cannot access the PowerShell credential cache (common on some machines), switch to `AzureCliCredential`:
```json
"EnvironmentVariables": {
    "AZURE_TOKEN_CREDENTIALS": "AzureCliCredential"
}
```
Ensure `az login --tenant <tenant-id>` is active. If recording fails with credential errors from the subprocess, this is the likely fix.

#### Redeploy if resource group is missing

Test resource groups are auto-deleted after 12 hours. If tests fail with `ResourceGroupNotFound`, redeploy:
```powershell
./eng/scripts/Deploy-TestResources.ps1 -Paths {Toolset}
```

### 3d. Test Project Configuration (Critical)

The test `.csproj` **must** have these specific settings or tests will fail with "azmcp.exe not found":

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net10.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <IsPackable>false</IsPackable>
    <IsTestProject>true</IsTestProject>

>     <OutputType>Exe</OutputType>

Reviewer: Need to also include the `<HasLiveTests>` and `<HasUnitTests>` properties

Reviewer 2: ```suggestion
    <OutputType>Exe</OutputType>
    <HasLiveTests>true</HasLiveTests>
    <HasUnitTests>true</HasUnitTests>
```

---

> name: add-mcp-tools

Reviewer 2: ```suggestion
name: add-azure-mcp-tools
```

| **test-resources-post.ps1** | ✅ Required (even if basic) | ❌ Skip |
| **RBAC role assignments** | ✅ Required | ❌ Skip |
| **Live tests** | ✅ Required (recorded) | ❌ Skip |
| **Unit tests** | ✅ Required | ✅ Required |

---

## Phase 0: New Toolset Setup (skip if adding to existing toolset)

Create the toolset directory structure:

```
tools/Azure.Mcp.Tools.{Toolset}/
├── src/
│   ├── Azure.Mcp.Tools.{Toolset}.csproj
│   ├── {Toolset}Setup.cs
│   ├── Commands/
│   │   ├── {Resource}/
│   │   │   └── {Resource}{Operation}Command.cs
│   │   └── {Toolset}JsonContext.cs
│   ├── Options/
│   │   └── {Resource}/
│   │       └── {Resource}{Operation}Options.cs
│   ├── Services/
│   │   ├── I{Toolset}Service.cs
│   │   └── {Toolset}Service.cs
│   └── Models/
└── tests/
    ├── Azure.Mcp.Tools.{Toolset}.Tests/
    │   └── Azure.Mcp.Tools.{Toolset}.Tests.csproj
    ├── test-resources.bicep          (Azure service commands only)
    └── test-resources-post.ps1       (Azure service commands only)
```

Required setup steps:

1. Add package version to `Directory.Packages.props` (if Azure SDK needed)
 2. Register the project in solution files by running:

>     `powershell eng/scripts/Update-Solutions.ps1 -Al`

Reviewer: ```suggestion
    `pwsh eng/scripts/Update-Solutions.ps1 -All`
```

PowerShell is invoked using `pwsh` and missing a letter in `All`

    public string? MyOption { get; set; }

    [Option(OptionDescriptions.ResourceGroup)]
    public string? ResourceGroup { get; set; }

    [Option(OptionDescriptions.Subscription)]
    public string? Subscription { get; set; }

    [Option(OptionDescriptions.Tenant)]
    public string? Tenant { get; set; }

    [Option(Name = "retry")]
    public RetryPolicyOptions? RetryPolicy { get; set; }
}
```

Rules:
- Implement `ISubscriptionOption` for commands that need subscription resolution
- Use `[Option("description")]` for the description — property name auto-converts to `--kebab-case`
- Use `[Option(Name = "custom")]` only when the default kebab-case conversion is wrong (e.g., `RetryPolicy` → `--retry` not `--retry-policy`)
- Use `[Option(OptionDescriptions.X)]` for shared descriptions (`Subscription`, `Tenant`, `ResourceGroup`, `AuthMethod`)
- Use `subscription` (never `subscriptionId`) — supports both IDs and names
- Use `resourceGroup` (never `resourceGroupName`)
- Use singular nouns for resources (`server` not `serverName`)
- Remove unnecessary `name` suffixes (`Account` / `--account` not `AccountName` / `--account-name`)

> - Nullability determines required/optional for reference types: `string?` = optional, `required string` = required

Reviewer: I updated this in a new PR as the `required` keyword determines nullability.

```suggestion
- Use `required` on required options; use nullable types (`?`) for optional options.
```
